llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp

```diff
@@ -623,13 +623,33 @@ void NVPTXLowerArgs::handleByValParam(const NVPTXTargetMachine &TM,
     Value *ArgInParam = new AddrSpaceCastInst(
         Arg, PointerType::get(Arg->getContext(), ADDRESS_SPACE_PARAM),
         Arg->getName(), FirstInst);
+    // Create an opaque type of same size as StructType but without padding
+    // holes as this could have been a union.
+    const auto StructBytes = *AllocA->getAllocationSize(DL);
+    SmallVector<Type *, 5> ChunkTypes;
+    if (StructBytes >= 16) {
+        Type *IntType = Type::getInt64Ty(Func->getContext());
+        Type *ChunkType = VectorType::get(IntType, 2, false);
+        Type *OpaqueType = StructBytes < 32 ? ChunkType :
+                           ArrayType::get(ChunkType, StructBytes / 16);
+        ChunkTypes.push_back(OpaqueType);
+    }
+    for (const auto ChunkBytes: {8, 4, 2, 1}) {
+      if (StructBytes & ChunkBytes) {
+          Type *ChunkType = Type::getIntNTy(Func->getContext(), 8 * ChunkBytes);
+          ChunkTypes.push_back(ChunkType);
+      }
+    }
+    Type * OpaqueType = ChunkTypes.size() == 1 ? ChunkTypes[0] :
+                        StructType::create(ChunkTypes);
```

Comment:
@Artem-B, here is one: https://godbolt.org/z/7EKzMEeK3. That pointer to a `{ [45 x i8] }` has 16-byte alignment, so should vectorize into a `{ [2 x <2 x i64>], i64, i32, i8 }`. You can see that _load-store-vectorizer_ runs after _nvptx-lower-args_, but does nothing. Debug output shows messages like `LSV: Access of ... in addrspace 5 with alignment 1 is misaligned, and therefore can't be vectorized.`, which looks like some alignment is not preserved. Once you find a fix for that, I'll remove the manual vectorization from my pull request.

Reply:
Interesting. I think the problem is that `SROA` just sticks 'align 1' on all 1-byte loads/stores, because it technically does not need any higher alignment, and LSV takes it at face value, even though the base pointer is aligned.

That looks like something we may want to fix, eventually, but it's probably going to take some time. 

Can we just use memcpy instead of load/store w/ a custom type? That should copy the data we need *and* convey proper alignment info to the code that does the copying.

Reply:
> Can we just use memcpy instead of load/store w/ a custom type?

Yes, we should be able to do that, I just couldn’t figure out how.

Reply:
Here's an example of creating an intrinsic call: https://github.com/llvm/llvm-project/blob/fc4b1a303b296d02f6243a083510c4ee7f290ab0/llvm/lib/CodeGen/IntrinsicLowering.cpp#L465
Memcpy should be similar, though it may need extra parameters to get the right intrinsic declaration as they are overloaded on pointers' AS.

Reply:
Thanks @Artem-B, that actually turned out to be quite easy. Latest commit switches to memcpy and updates tests accordingly.